community/dbms/src/main/java/org/neo4j/commandline/dbms/MemoryRecommendationsCommand.java

```diff
@@ -0,0 +1,385 @@
+            outputUpdatedConf( confFile, heap, pagecache, overwrite, updatedLines::add );
+            try
+            {
+                Files.write( confFile, updatedLines, StandardCharsets.UTF_8 );
+                outsideWorld.stdErrLine( "File updated: " + confFile );
+            }
+            catch ( IOException e )
+            {
+                throw new CommandFailed( "Failed to update configuration file: " + confFile, e );
+            }
+        }
+    }
+
+    private void outputUpdatedConf(
+            Path confFile, String heap, String pagecache, boolean overwrite, Consumer<String> out ) throws CommandFailed
+    {
+        try
+        {
+            // Existing settings to maybe skip.
+            Map<String,String> existingSettings = MapUtil.load( confFile.toFile() );
+            // Settings that were updated by our pass-through.
+            Set<String> redefined = new HashSet<>();
+            // Skip settings that are either already defined in the conf file, or were defined earlier in our pass.
+            Predicate<String> skip = s -> existingSettings.containsKey( s ) || redefined.contains( s );
+            try ( Stream<String> lines = Files.lines( confFile, StandardCharsets.UTF_8 ) )
+            {
+                lines.forEachOrdered( line ->
+                {
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, initialHeapSize.name(), heap );
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, maxHeapSize.name(), heap );
+                    line = tryUpdateSetting( line, skip, redefined, overwrite, pagecache_memory.name(), pagecache );
+                    out.accept( line );
+                } );
+            }
+            defineIfMissing( out, redefined, initialHeapSize.name(), heap );
+            defineIfMissing( out, redefined, maxHeapSize.name(), heap );
+            defineIfMissing( out, redefined, pagecache_memory.name(), pagecache );
+        }
+        catch ( Exception e ) // Also catch runtime exceptions due to use of Streams.
+        {
+            throw new CommandFailed( "Failed to read configuration file: " + confFile, e );
+        }
+    }
+
+    private String tryUpdateSetting( String line, Predicate<String> alreadyDefined, Set<String> redefined,
+                                     boolean overwrite, String name, String value )
+    {
+        int settingIndex = line.indexOf( name );
+        if ( settingIndex == -1 )
+        {
+            // The setting we're looking for is not present in this line, so we just return it unmodified.
+            return line;
+        }
+        // Otherwise, we need to modify this line, such that it reflects the new value.
+        // Except, if 'overwrite' is false, then we don't modify the line if the setting is not commented out.
+        int firstComment = line.indexOf( '#' );
+        if ( !overwrite && (settingIndex < firstComment || firstComment == -1 || alreadyDefined.test( name )) )
```

Comment:
I'm confused about `alreadyDefined` in this statement. Why is it needed? And if it is needed, then why is it `OR`d with `!overwrite`?